Move contacts to use nock

MattMSumner opened the pull request:
Why:

The contacts test failed on CI due to a contact no longer existing
https://travis-ci.org/MadKudu/node-hubspot/builds/469517133

This PR:

Moves the contacts tests to use nock and mock out the api calls.

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
 
   describe('getRecentlyCreated', function() {
-    it('should return last 100 updated contacts', function() {
+    const createdContactsGetEndpoint = {
```
Small naming suggestion: `recentlyCreatedContactsGetEndpoint`.

MattMSumner replied:
Should I also change the above one to `recentlyModifiedContactsGetEndpoint`?

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
 
   describe('getRecentlyModified', function() {
-    it('should return last 100 updated contacts', function() {
+    const modifiedContactsGetEndpoint = {
```
Small naming suggestion: `recentlyModifiedContactsGetEndpoint`.

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
-      })
+      if (process.env.NOCK_OFF) {
+        return hubspot.contacts.get().then(data => {
+          contactId = data.contacts[0].vid
```
Will the call to `.vid` fail if no contacts exist? Is it reasonable to check if the return value is an array? This way, the avoids caring about data that may or may not exist in the HubSpot app.

MattMSumner replied:
I changed this to create and delete a contact in 
https://github.com/MadKudu/node-hubspot/pull/140/commits/d29edc475dc673b6e931ac649c1f0f82f31a3b4c

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
   describe('getByIdBatch', function() {
-    let contactIds
+    let contactIds = [123, 234, 345]
+    const getContactByIdsEndpoint = {
```
small naming suggestion: `getContactsByIdsEndpoint`

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
   describe('getById', function() {
-    let contactId
+    let contactId = 123
+    const getContactByIdEndpoint = {
```
There's a slight variation in naming that's probably worth synchronizing. For example, the tests above place the HTTP verb before `Endpoint` (`createdContactsGetEndpoint`) and here the verb is first. Thoughts on using just one format? FWIW, I like the format that places the HTTP verb first: `getContactByIdEndpoint`

MattMSumner replied:
I went the other way and removed all the gets in 
https://github.com/MadKudu/node-hubspot/pull/140/commits/934f677dee32f0d73f0dc4c3efb99fbf33326704

SViccari replied:
I like it 👍

SViccari commented on test/contacts.js:
```diff
@@ -1,14 +1,35 @@
+    const tenContactsGetEndpoint = {
+      path: '/contacts/v1/lists/all/contacts/all',
+      response: { contacts: [{}, {}, {}, {}, {}, {}, {}, {}, {}, {}] },
+      query: { count: 10 },
```
Does this work similar to an expectation where calling `hubspot.contacts.get({ count: 5 })` would cause this test to fail because the count didn't match the mocked endpoint? 

If that's true, what do you think of removing the assertion: `expect(data.contacts.length).to.be.at.most(10)`? 

I believe this expectation is based on the set response, which isn't very helpful or informative (IMO). I think the real value is asserting the correct endpoint is called, with the correct params, and the response matches an expected object (in this case, an array). Thoughts?

MattMSumner replied:
done in 
https://github.com/MadKudu/node-hubspot/pull/140/commits/c3ec3aed32e3df408aba03a6737d3f0cb740a4b3

SViccari commented on test/contacts.js:
```diff
@@ -17,56 +38,83 @@ describe('contacts', function() {
-      })
+      if (process.env.NOCK_OFF) {
+        return hubspot.contacts.get({ count: 3 }).then(data => {
+          contactIds = _.map(data.contacts, 'vid')
```
Similar question here, will this fail if there are only 2 contacts?

MattMSumner replied:
I think this would be fine with only 2 contacts since we map over the result.

SViccari commented on test/contacts.js:
```diff
@@ -78,172 +126,232 @@ describe('contacts', function() {
+
+    before(function() {
+      if (process.env.NOCK_OFF) {
+        return hubspot.contacts.get({ count: 3 }).then(data => {
```
Is there a benefit to setting the count to 3 instead of requesting one? It seems we just need the API to return one contact so the contact's email can be used in the test (when nock is off)

MattMSumner replied:
I guess to make sure we get one with an email? This is defensive so I can change this to 1 👍

SViccari replied:
Ah, is email not a required field on Contact?

SViccari commented on test/contacts.js:
```diff
@@ -78,172 +126,232 @@ describe('contacts', function() {
+
+    it('should return a contact record based on a array of emails', function() {
+      return hubspot.contacts.getByEmailBatch(emails).then(data => {
+        expect(data).to.be.an('object')
```
I think it's wise to avoid testing too much about the return response since that logic lives with the HubSpot API but is there anything more meaningful we can assert for here? Perhaps one key/field that is relevant to a Contact?

MattMSumner replied:
The response from HubSpot is an object with the keys set to the ids of the records. Because I won't know what these are beforehand, I don't think I should make any assertions beyond that it succeeds and I get back an object.

SViccari commented on test/contacts.js:
```diff
@@ -78,172 +156,232 @@ describe('contacts', function() {
+        {
+          property: 'lastname',
+          value: 'Lead',
+          timestamp: Date.now(),
```
Just curious, is the `timestamp` necessary? I'm also interested in the decision to update these particular fields. Can this test update fewer fields and still confirm `update` works as expected?

MattMSumner replied:
I removed them and it seems fine 
https://github.com/MadKudu/node-hubspot/pull/140/commits/2026e197fbbbb2b6e152cf10f74af9559ceaa224

SViccari commented on test/contacts.js:
```diff
@@ -1,14 +1,61 @@
+  const hubspot = new Hubspot({
+    accessToken: process.env.ACCESS_TOKEN || 'fake-token',
+  })
+  const createTempContact = () =>
```
This is a bit nit-picky but what do you think of re-naming `createTempContact()` to `createTestContact()`? There's nothing that makes the contact temporary but I think the purpose is to communicate that this contact doesn't matter past the context of this test.

SViccari reviewed (approved):
Left some comments but overall, LGTM 🐙

MattMSumner committed:
Move contacts to use nock

Why:

The contacts test failed on CI due to a contact no longer existing
https://travis-ci.org/MadKudu/node-hubspot/builds/469517133

This PR:

Moves the contacts tests to use nock and mock out the api calls.
